From 89ede6daa709192801c4b59830112a407e1ca54f Mon Sep 17 00:00:00 2001
From: Jean-Yves Avenard <jyavenard@mythtv.org>
Date: Sat, 31 May 2014 01:59:28 +1000
Subject: [PATCH 17/28] Update AVFormatWriter for new FFMpeg APIs

Fix a memory leak, object m_picture allocated was never freed and remove unused member
---
 mythtv/libs/libmythtv/avformatwriter.cpp |   24 +++++++++++-------------
 mythtv/libs/libmythtv/avformatwriter.h   |    1 -
 2 files changed, 11 insertions(+), 14 deletions(-)

diff --git a/mythtv/libs/libmythtv/avformatwriter.cpp b/mythtv/libs/libmythtv/avformatwriter.cpp
index 32f321e..900fcd6 100644
--- a/mythtv/libs/libmythtv/avformatwriter.cpp
+++ b/mythtv/libs/libmythtv/avformatwriter.cpp
@@ -45,7 +45,7 @@ AVFormatWriter::AVFormatWriter()
       m_ctx(NULL),
       m_videoStream(NULL),   m_avVideoCodec(NULL),
       m_audioStream(NULL),   m_avAudioCodec(NULL),
-      m_picture(NULL),       m_tmpPicture(NULL),
+      m_picture(NULL),
       m_audPicture(NULL),
       m_audioInBuf(NULL),    m_audioInPBuf(NULL)
 {
@@ -78,6 +78,8 @@ AVFormatWriter::~AVFormatWriter()
         avcodec_free_frame(&m_audPicture);
 
     Cleanup();
+
+    av_frame_free(&m_picture);
 }
 
 bool AVFormatWriter::Init(void)
@@ -244,6 +246,7 @@ int AVFormatWriter::WriteVideoFrame(VideoFrame *frame)
     planes[2] = planes[1] + (frame->width * frame->height) /
         4; // (pictureFormat == PIX_FMT_YUV422P ? 2 : 4);
 
+    av_frame_unref(m_picture);
     m_picture->data[0] = planes[0];
     m_picture->data[1] = planes[1];
     m_picture->data[2] = planes[2];
@@ -583,25 +586,20 @@ bool AVFormatWriter::OpenVideo(void)
         return false;
     }
 
-    m_picture = AllocPicture(c->pix_fmt);
     if (!m_picture)
     {
-        LOG(VB_RECORD, LOG_ERR,
-            LOC + "OpenVideo(): AllocPicture() failed");
-        return false;
-    }
-
-    m_tmpPicture = NULL;
-    if (c->pix_fmt != PIX_FMT_YUV420P)
-    {
-        m_tmpPicture = AllocPicture(PIX_FMT_YUV420P);
-        if (!m_tmpPicture)
+        m_picture = AllocPicture(c->pix_fmt);
+        if (!m_picture)
         {
             LOG(VB_RECORD, LOG_ERR,
-                LOC + "OpenVideo(): m_tmpPicture AllocPicture() failed");
+                LOC + "OpenVideo(): AllocPicture() failed");
             return false;
         }
     }
+    else
+    {
+        av_frame_unref(m_picture);
+    }
 
     return true;
 }
diff --git a/mythtv/libs/libmythtv/avformatwriter.h b/mythtv/libs/libmythtv/avformatwriter.h
index 6f0be09..aeb4a9a 100644
--- a/mythtv/libs/libmythtv/avformatwriter.h
+++ b/mythtv/libs/libmythtv/avformatwriter.h
@@ -52,7 +52,6 @@ class MTV_PUBLIC AVFormatWriter : public FileWriterBase
     AVStream              *m_audioStream;
     AVCodec               *m_avAudioCodec;
     AVFrame               *m_picture;
-    AVFrame               *m_tmpPicture;
     AVFrame               *m_audPicture;
     unsigned char         *m_audioInBuf;
     unsigned char         *m_audioInPBuf;
-- 
1.7.10.2

